File: binary_search/binarySearch.js
const bisectSearch = (arr, target, lo = 0, hi = null) => {
    if (lo < 0) {
        throw new Error('lo must be non-negative');
    }
    if (hi === null) {
        hi = arr.length;
    }
    while (lo < hi) {
        let mid = Math.floor((lo + hi)/2);
        if (arr[mid] < target) {
            lo = mid + 1
        } else {
            hi = mid
        }
    }
    return lo;
}

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
const search = (nums, target) => {
    const index = bisectSearch(nums, target);
    return nums[index] === target ? index : -1;
};

File: binary_search/binary_search.py
# https://leetcode.com/problems/binary-search/

def bisect_left(a, x, lo=0, hi=None):
    if lo < 0:
        raise ValueError('lo must be non-negative')
    if hi is None:
        hi = len(a)
    while lo < hi:
        mid = (lo+hi)//2
        if a[mid] < x: lo = mid+1
        else: hi = mid
    return lo

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        i = bisect_left(nums, target)
        if i != len(nums) and nums[i] == target:
            return i
        else:
            return -1
